Clarify explainer and address some less-complex issues (#36)
jackbsteinberg/std-toast: README.md

  either by stacking them in the view,
  or queueing them and displaying sequentially.

> A broad goal of the standard toast API is to provide a base for more opinionated or featureful toast libraries to layer on top of.

Reviewer: Delete? To me, this just makes it sound like it might not really be a goal!

Author: I changed the language so it feels like more of a priority. Providing layering for libraries is important as a goal so I don't think it should be taken out, but instead emphasized differently.

The element is provided as a [built-in module](https://github.com/tc39/proposal-javascript-standard-library/blob/master/README.md),
named `"std:elements/toast"`.

> Read more about "pay-for-what-you-use" HTML elements [here](https://github.com/whatwg/html/issues/4697).

Reviewer 2: ```suggestion
See [whatwg/html#4697](https://github.com/whatwg/html/issues/4697) for more discussion on "pay-for-what-you-use" elements available via module imports.
```

to give a bit more context and useful linking text.

### The `<std-toast>` element

#### Attributes

- [Global attributes](https://html.spec.whatwg.org/multipage/dom.html#global-attributes)
- `open`: a boolean attribute, determining whether the toast is visible or not (according to the default styles).
By default toasts are not shown.
- `theme`: one of `"default"`, ???, or ???, conveying the semantic priority of the toast, and influencing its styling (both default and user-provided)
    - TODO([#18](https://github.com/jackbsteinberg/std-toast/issues/18)): decide on list of themes to natively support and create and style them.
- `position`: default position will be ???
    - Options for position:
        - `"top-left"`
        - `"top-center"`
        - `"top-right"`
        - `"middle-left"`
        - `"middle-center"`
        - `"middle-right"`
        - `"bottom-left"`
        - `"bottom-center"`
        - `"bottom-right"`
    - Additionally values `"top-stretch"` and `"bottom-stretch"` will be provided to allow for toasts that span the screen width,

>     and will be treated as the default for `"top-"` and `"bottom-"` toasts on mobile.

Reviewer 2: I'm not sure we should be definitive about this particular API choice yet. Maybe instead some kind of TODO about stretch styles and about mobile vs. desktop?

Author: Created issue #39 and referenced in commit

Author: I removed the `middle-left` and `middle-right` positions, and changed the center position to just be `center` by the way, they seemed like values we didn't need to include

- `.show(options)`: shows the toast element,
by toggling its `open=""` attribute to true.
It will also start or reset the timeout of the toast,
to show for the provided `duration`,
or a default `duration` of `2000`ms.
The `options` include:

>     - `duration`: how long to show the toast, in milliseconds. Defaults to `2000`.

Reviewer 2: Link to #37 here.

Author: Done.

- `.show(options)`: shows the toast element,
by toggling its `open=""` attribute to true.

> it will also start or reset the timeout of the toast,

Reviewer: ```suggestion
It will also start or reset the timeout of the toast,
```

> to show fo the provided `duration`,

Reviewer 2: ```suggestion
to show for the provided `duration`,
```

> to show fo the provided `duration`,

Reviewer: ```suggestion
to show for the provided `duration`,
```

or a default `duration` of `2000`ms.
The `options` include:

>     - `duration`: how long to show the toast, in milliseconds. Defaults to `2000`.

Reviewer 2: I think we should file an issue for the default timeout. It should include research on all the libraries, plus discussion about accessibility. E.g. some of the top Google search results for accessible toast suggest heuristics like 3 seconds + 60 milliseconds per character or similar.

Author: #37

a CSS shadow part for the close button,
or some CSS variables.

## Accessibility

> 

Reviewer 2: I'd add something like "TODO: determine detailed accessibility semantics; see #25 for initial discussion."

Author: Done.

> or like in GMail ("Chat is currently unavailable") etc. 
> These are not things you want an actual system notification for... 
> They really depend on the surrounding content/situation unlike regular notifications.

The toast is intended to be a completely in-page element,
not a system-level notification the way Notifications and Android toasts are.
Because of this page-level role, toasts do not require any user-granted permissions.

### Extending the `<dialog>` element
This approach was considered, and is still an open area of exploration,
but our current thoughts are to go a different route for a few reasons.
First and foremost, we felt the accessibility considerations for a toast and a dialog differ,
as the dialog typically engages in the kind of flow-breaking behavior we want toasts to avoid.
Additionally, the popular pattern of adding a call to action button on a toast with a baked-in time limit
necessitates replacing tab-trapping with a less intrusive, easily restored alternative.
The line between the elements becomes blurrier from a functional perspective,
with action-button-containing toasts and non-modal dialogs,
but the intent and semantics are sufficiently different,

> similar to `<blockquote>` and `<q>` and `<meter>` and `<progress>`.

Reviewer 2: ```suggestion
similar to `<blockquote>` and `<q>`, or `<meter>` and `<progress>`.
```

```

### Styling the toast
TODO

## Security and Privacy Considerations
See [TAG Security / Privacy Self Review](/security-privacy-self-review.md).

## Considered Alternatives

> 

Reviewer: Accessibility section? Even it's also empty.

Author: I'm actually working on an accessibility study in another branch, I can add the header in here though

### Extending the Notification API
Toasts are intended to be ephemeral, context-specific messages,
and collecting them in a user's notifications tray doesn't reflect the spirit of the pattern.

> [kenchris]() put this well in [this comment]() on the TAG Design Review:

Reviewer 2: Missing links?

Author: ```suggestion
[kenchris](https://github.com/kenchris) put this well in [this comment](https://github.com/w3ctag/design-reviews/issues/385#issuecomment-502070938) on the TAG Design Review:
```

> the difference is that this is a temporary notification / infobar, 
> like you see for PWAs ("new version available, press reload to refresh") 
> or like in GMail ("Chat is currently unavailable") etc. 
> These are not things you want an actual system notification for... 
> They really depend on the surrounding content/situation unlike regular notifications.

Toasts also often contain some sort of action the user can take which affects the state of the page 
(e.g. un-sending an email on Gmail).

> To provide this kind of behavior accessibly the element should live wholly in the page / DOM.

Reviewer 2: I'm not sure these two sentences are accurate. Notifications can have actions these days, and I assume the OS vendors have made their notifications accessible.

Author: Yeah I had the wrong impression initially, updated to reflect.

> This will have the added benefit of preventing pages in the background from using toasts to fight for user attention.

Reviewer 2: Maybe also point out that because toasts are in-page only and non-intrusive in this way, they do not require user permission, like notifications do.

Author: Done.

### Extending the `<dialog>` element
This approach was considered, but we decided to go a different route for a few reasons.
First and foremost, we felt the accessibility considerations for a toast and a dialog are fundamentally opposed,
as the dialog engages in the kind of flow-breaking behavior we want toasts to avoid.
Additionally, the popular pattern of adding a call to action button on a toast with a baked-in time limit

> necessitates replacing tab-trapping with a less intrusive, easily restored alternative.

Reviewer 2: Maybe we should soften this a bit by talking about how the exact differences between them are still being explored, and especially for action-button-containing toasts, the line might be a bit blurry. Then we could talk about how even if the line is blurry functionally, it's very different in terms of intent and semantics, similar to blockquote and q, or meter and progress.

Author: Done, hope it's more clear now.

### As a new global element (instead of a built-in module)

> See [here](https://github.com/tkent-google/std-switch#as-a-new-global-element-instead-of-a-built-in-module).

Reviewer 2: Let's instead link to https://github.com/whatwg/html/issues/4697

Author: ```suggestion
See [here](https://github.com/whatwg/html/issues/4697).
```

### Leaving this up to libraries
See [here](https://github.com/tkent-google/std-switch#leaving-this-up-to-libraries).

> ## Extra Resources

Reviewer: discourse thread? I2I thread?

Author: Added the discourse thread

> See [here](https://github.com/tkent-google/std-switch#faqs) for std-switch FAQs,

Reviewer 2: Link text should generally not be "here", e.g. "See [the std-switch FAQs]", or use the issue titles for the whatwg/html issues. I think you should also list the three FAQ titles so people get a sense (before clicking) of what questions they're being referred to.

Author: Ah I'll go in and apply some context to all of these, my bad

and [here](https://github.com/whatwg/html/issues/4696)
and [here](https://github.com/whatwg/html/issues/4697)

> for relevant open issues on WHATWG/html.

Reviewer 2: ```suggestion
for relevant open issues on the HTML Standard regarding potential patterns for new built-in elements (including this one).
```
